Userland/Libraries/LibGLSL/Lexer.cpp

```diff
@@ -0,0 +1,829 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2023, Volodymyr V. <vvmposeydon@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Lexer.h"
+#include <AK/CharacterTypes.h>
+#include <AK/DeprecatedString.h>
+#include <AK/Function.h>
+#include <AK/HashTable.h>
+#include <AK/StdLibExtras.h>
+
+namespace GLSL {
+
+Lexer::Lexer(StringView input, size_t start_line)
+    : m_input(input)
+    , m_previous_position { start_line, 0 }
+    , m_position { start_line, 0 }
+{
+}
+
+char Lexer::peek(size_t offset) const
+{
+    if ((m_index + offset) >= m_input.length())
+        return 0;
+    return m_input[m_index + offset];
+}
+
+char Lexer::consume()
+{
+    VERIFY(m_index < m_input.length());
+    char ch = m_input[m_index++];
+    m_previous_position = m_position;
+    if (ch == '\n') {
+        m_position.line++;
+        m_position.column = 0;
+    } else {
+        m_position.column++;
+    }
+    return ch;
+}
+
+constexpr bool is_valid_first_character_of_identifier(char ch)
+{
+    return is_ascii_alpha(ch) || ch == '_' || ch == '$';
+}
+
+constexpr bool is_valid_nonfirst_character_of_identifier(char ch)
+{
+    return is_valid_first_character_of_identifier(ch) || is_ascii_digit(ch);
+}
+
+// NOTE: some of these keywords are not used at the moment, however they are reserved for future use and should not be used as identifiers
+constexpr StringView s_known_keywords[] = {
+    "asm"sv,
+    "attribute"sv,
+    "break"sv,
+    "case"sv,
+    "cast"sv,
+    "centroid"sv,
+    "class"sv,
+    "common"
+    "partition"sv,
+    "active"sv,
+    "const"sv,
+    "continue"sv,
+    "default"sv,
+    "discard"sv,
+    "do"sv,
+    "else"sv,
+    "enum"sv,
+    "extern"sv,
+    "external"sv,
+    "false"sv,
+    "filter"sv,
+    "fixed"sv,
+    "flat"sv,
+    "for"sv,
+    "goto"sv,
```

Comment:
Please double check which of these keywords are actually valid in glsl. I'm pretty sure that goto, for example, is not valid glsl.

Reply:
Some of those keywords are not used in the current version of GLSL, however per chapter 3.6 of the GLSL specification they are still reserved for future use and cannot be used for naming user-defined objects.

Reply:
Ok then please forget what I said :^)